service/history/hsm/tree.go

```diff
@@ -700,6 +725,44 @@ func (n *Node) root() *Node {
 	return root
 }
 
+// Compact filters the operation log to remove any transitions for deleted nodes while preserving deletion operations.
+// The returned log maintains deletion tombstones, but excludes transitions that occurred on deleted nodes or their
+// descendants.
+func (ol OperationLog) Compact() OperationLog {
+	var deletedPaths [][]Key
+
+	// First collect all deletions
+	for _, op := range ol {
+		if del, ok := op.(DeleteOperation); ok {
+			deletedPaths = append(deletedPaths, del.Path())
+		}
+	}
+
+	// Then filter transitions
+	result := make(OperationLog, 0, len(ol))
+	for _, op := range ol {
+		switch o := op.(type) {
+		case DeleteOperation:
+			// Keep all delete operations as tombstones
```

Comment:
If the parent is deleted, you want to compact the child deletion away.

Reply:
Done!